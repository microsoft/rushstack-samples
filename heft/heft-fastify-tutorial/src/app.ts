import * as path from 'path';
import * as ejs from 'ejs';
import { FastifyPluginAsync, FastifyInstance } from 'fastify';
import PointOfView from 'point-of-view';
import FastifyStatic from 'fastify-static';

import { HomeController } from './web/HomeController';
import { BooksController } from './web/books/BooksController';
import { ApiBooksController } from './api/books/ApiBooksController';

const app: FastifyPluginAsync = async (fastify: FastifyInstance) => {
  // Register plugins
  await fastify.register(PointOfView, {
    engine: {
      ejs: ejs
    },
    layout: 'layouts/global.ejs',
    root: path.join(__dirname, './web')
  });
  await fastify.register(FastifyStatic, {
    root: path.join(__dirname, './assets'),
    prefix: '/assets'
  });

  // Application routes, organized into small "controllers"
  await fastify.register(HomeController.register, { prefix: '/' });
  await fastify.register(BooksController.register, { prefix: '/books' });
  await fastify.register(ApiBooksController.register, { prefix: '/api/books' });
};

// Our `app.ts` exports a Fastify Plugin, which means you can run it directly
// with the fastify-cli after compiling:
//
//   fastify lib/app.js
//
// Doing this also keeps unit tests clean, as they can import the app definition
// without getting in the mud with the specifics of starting the server.
//
// In production, you'll want to be able to configure Fastify (for example,
// to specify where output logs should go), so we would run `server.ts` instead:
//
//   node lib/server.js
//
export default app;
