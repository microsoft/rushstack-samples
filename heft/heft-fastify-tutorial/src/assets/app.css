body,
html {
  background-color: #be5f3e;
  font-family: Helvetica, Verdana, Arial, sans-serif;
  margin: 0px;
  padding: 0px;
}

.site {
  background-color: #fffff0;
  border-radius: 8px;
  color: #333;
  padding: 16px;
  margin: 16px;
}

h2 {
  margin-top: 0px;
  border-top: 0px;
}

table.books {
  border-collapse: collapse;
}

table.books td,
table.books th {
  border: solid 1px #666;
  padding: 0.25em 0.5em;
}
