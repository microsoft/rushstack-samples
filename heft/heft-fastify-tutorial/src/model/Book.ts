export class Book {
  public id: number | undefined;
  public title: string;
  public author: string;

  public constructor(title: string, author: string) {
    this.title = title;
    this.author = author;
  }

  // In real life, Book would be an interface to MongoDB, or DynamoDB,
  // or a relational database of some kind. For this simple example,
  // we'll return a fake static list of books.
  public static list(): Book[] {
    return [
      new Book('Moby Dick', 'Herman Melville'),
      new Book('Rabbit Island', 'Elvira Navarro'),
      new Book('Accelerando', 'Charles Stross')
    ];
  }
}
