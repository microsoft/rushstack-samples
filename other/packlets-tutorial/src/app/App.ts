// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { DataModel, ExampleModel } from '../packlets/data-model';

import { Logger, MessageType } from '../packlets/logging';
import { MainReport } from '../packlets/reports';

export class App {
  public run(): void {
    const logger: Logger = new Logger();
    logger.log(MessageType.Info, 'Starting app...');

    const dataModel: DataModel = new ExampleModel(logger);
    const report: MainReport = new MainReport(logger);
    report.showReport(dataModel);

    logger.log(MessageType.Info, 'Operation completed successfully');
  }
}
