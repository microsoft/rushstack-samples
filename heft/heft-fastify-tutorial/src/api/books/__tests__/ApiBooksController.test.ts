import fastify from 'fastify';
import app from '../../../app';
import { Book } from '../../../model/Book';

describe('ApiBooksController', () => {
  describe('list', () => {
    it('renders JSON view with a list of books', async () => {
      jest.spyOn(Book, 'list').mockReturnValue([new Book('A test book', 'A test author')]);

      const subject = fastify().register(app);
      const response = await subject.inject({
        method: 'GET',
        url: '/api/books'
      });

      expect(response.statusCode).toEqual(200);
      expect(response.body).toMatchSnapshot();
    });
  });
});
