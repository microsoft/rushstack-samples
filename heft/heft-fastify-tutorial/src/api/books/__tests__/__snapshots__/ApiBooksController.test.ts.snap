// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiBooksController list renders JSON view with a list of books 1`] = `"{\\"books\\":[{\\"title\\":\\"A test book\\",\\"author\\":\\"A test author\\"}]}"`;
